src/utilities/update.ts

```diff
@@ -39,69 +39,145 @@ function projectValuesOntoExistingObj(dst: Record<string, unknown>, src: Record<
   return dst
 }
 
-/*
-  Return first index found from given array that is not an empty entry (cell)
-*/
-function getIndexOfFirstLine(auditArray: string[], index: number, action: string): number {
-  let indexVal = index;
-  while (auditArray[indexVal] === '') {
-    switch (action) {
-      case '-':
-        indexVal--
-        break;
-      case '+':
-        indexVal++
-        break;
+function getRangesFromStackUpdate(text: string): number[][] {  
+  const delims: {[key: string]: string} = {'(': ')', '{': '}', '[': ']', '<': '>'}
+  const quotes = '\'"`'
+  const strings = 'qQriIwWxs'
+
+  const stack: string[] = []
+  const rangeStack: number[][] = []
+  const ranges: number[][] = []
+
+  const lines = text.split('\n')
+  for (let i = 0; i < lines.length; i++) {
+    let j = 0
+    while (j < lines[i].length) {
+      const line = lines[i]
+      let char = line[j]
+
+      const isEmptyStack = (stack.length == 0)
+      const isNotEmptyStack = (stack.length > 0)
+
+      const isQuoteChar = quotes.includes(char)
+      const isNotEscapeChar = ((j == 0) || (j > 0 && line[j - 1] != '\\'))
+      const isPercentChar = (char == '%')
+      const isStringDelimiterChar = ((j < line.length - 1) && (/^[^A-Za-z0-9]$/.test(line[j + 1])))
+      const isCommentBeginChar = ((j == 0) && (line.length >= 6) && (line.slice(0, 6) == '=begin'))
+      
+      const conditionOne = ((j < line.length - 1) && (strings.includes(line[j + 1])))
+      const conditionTwo = ((j < line.length - 2) && (/^[^A-Za-z0-9]$/.test(line[j + 2])))
```

Comment:
State why we have to use -1 or -2

Reply:
We're checking for `j < line.length - 1` since we're accessing `line[j + 1]` in the next condition. In Python, this would throw an `IndexError`. In TypeScript, we seem to get an `undefined`, which could result in unexpected behavior.